app/gui2/shared/ast/tree.ts

```diff
@@ -306,6 +337,130 @@ export abstract class MutableAst extends Ast {
   }
 }
 
+/** Values that may be found in fields of `Ast` subtypes. */
+type FieldData =
+  | NodeChild<AstId>
+  | NodeChild
+  | NodeChild<SyncTokenId>
+  | FieldData[]
+  | undefined
+  | StructuralField
+/** Objects that do not directly contain `AstId`s or `SyncTokenId`s, but may have `NodeChild` fields. */
+type StructuralField =
+  | RawMultiSegmentAppSegment
+  | RawBlockLine
+  | RawOpenCloseTokens
+  | RawNameSpecification
+/** Type whose fields are all suitable for storage as `Ast` fields. */
+interface FieldObject {
+  [field: string]: FieldData
+}
+/** Returns the fields of an `Ast` subtype that are not part of `AstFields`. */
+function* fieldDataEntries<Fields>(map: FixedMapView<Fields>) {
+  for (const entry of map.entries()) {
+    // All fields that are not from `AstFields` are `FieldData`.
+    if (!astFieldKeys.includes(entry[0] as any)) yield entry as [string, FieldData]
```

Comment:
Since the key type is checked here, the return type can be narrowed.

```suggestion
    if (!astFieldKeys.includes(entry[0] as any)) yield entry as [keyof RawAstFields, FieldData]
```

This should help further reduce casts down the line.

Also to contain those remaining (trivial in-context) type casts, a separate `isAstFieldKey` type asserting function could be created. Same for `FieldData` validation, though this one may be a bit harder to balance safety and runtime verification overhead. But given that this effectively deals with arbitrary network received data, we shouldn't just blindly trust it.

Reply:
This return the fields that are *not* from `AstFields`, so I don't see a good way to narrow it.

Runtime AST field validation would need to occur when applying updates to the module, and seems orthogonal to this new functionality--on `develop` we already trust the received AST fields not to cause TypeErrors.